proto/asn1/asn1-universal-types/asn1_universal_types_to_der.cc

```diff
@@ -0,0 +1,107 @@
+#include "asn1_universal_types_to_der.h"
+#include <google/protobuf/util/time_util.h>
+#include <time.h>
+
+namespace asn1_universal_types {
+
+void Encode(const BitString& bit_string, std::vector<uint8_t>& der) {
+  // BitString has tag number 3 and can be both constructed and primitive
+  // (X.208, Table 1).
+  // Add one to the length for the unused bits byte.
+  EncodeTagAndLength((bit_string.encoding() << 5 | 0x03),
+                     bit_string.val().size() + 1, der.size(), der);
+
+  // Encode 0 to indicate that there are no unused bits.
+  // This also acts as EOC if val is empty.
+  der.push_back(0x00);
+  der.insert(der.end(), bit_string.val().begin(), bit_string.val().end());
+}
+
+void Encode(const Integer& integer, std::vector<uint8_t>& der) {
+  // Integer has tag number 2 (X.208, Table 1) and is always primitive (X.690
+  // (2015), 8.3.1).
+  EncodeTagAndLength(0x02, integer.val().size(), der.size(), der);
+
+  der.insert(der.end(), integer.val().begin(), integer.val().end());
+}
+
+void Encode(const UTCTime& utc_time, std::vector<uint8_t>& der) {
+  const size_t size_before = der.size();
+
+  const auto time_stamp = utc_time.time_stamp();
+  time_t rawtime =
+      google::protobuf::util::TimeUtil::TimestampToTimeT(time_stamp);
+  struct tm* ptm;
+  ptm = gmtime(&rawtime);
+
+  const uint16_t years_since_1900 = ptm->tm_year;
+  der.push_back(0x30 + 0x01 + (years_since_1900 / 1000));
+  der.push_back(0x30 + 0x09 + (years_since_1900 / 100));
+  der.push_back(0x30 + (years_since_1900 / 10));
+  der.push_back(0x30 + (years_since_1900 % 10));
+
+  const uint8_t months_since_jan = ptm->tm_mon;
+  der.push_back(0x31 + (months_since_jan / 10));
+  der.push_back(0x31 + (months_since_jan % 10));
+
+  const uint8_t day_of_month = ptm->tm_mday;
+  der.push_back(0x30 + (day_of_month / 10));
+  der.push_back(0x30 + (day_of_month % 10));
+
+  const uint8_t hours_since_midnight = ptm->tm_hour;
+  der.push_back(0x30 + (hours_since_midnight / 10));
+  der.push_back(0x30 + (hours_since_midnight % 10));
+
+  const uint8_t minutes_after_hour = ptm->tm_min;
+  der.push_back(0x30 + (minutes_after_hour / 10));
+  der.push_back(0x30 + (minutes_after_hour % 10));
+
+  // The encoding shall terminate with "Z" (X.690 (2015), 11.7.1 | ISO/IEC
+  // 8824-1).
+  der.push_back(0x5a);
```

Comment:
This is on me, but I missed that `time_util.h` had enough of the Google date/time library to make this *much* easier for you.

```suggestion
std::string iso_date = google::protobuf::util::TimeUtil::ToString(time_stamp);
if (iso_date.size() <= 25)
  return;

string generalized_time;
generalized_time.reserve(15);
generalized_time += iso_date.substr(0, 4);  // Year
generalized_time += iso_date.substr(6, 2);  // Month
generalized_time += iso_date.substr(9, 2);  // Day
generalized_time += iso_date.substr(12, 2);  // Hour
generalized_time += iso_date.substr(15, 2);  // Minute
generalized_time += iso_date.substr(18, 2);  // Seconds
generalized_time += "Z";

der.resize(15);
for (const char c : generalized_time) {
  der.push_back(static_cast<uint8_t>(c));
}
```

I probably counted wrong for the offsets, but that at least gives you some handy tricks. Sadly, the function to convert to a `DateTime` is in the internal namespace (and thus you can't use), but that's what ToString is using and supports the range 0001 to 9999, which is what we want :)

Note the above is GeneralizedTime: UTCTime only supports the range 1950 - 2050 (because it uses 2 digits for the year, not 4).

Last Thursday, we talked about two options: either fixing up invalid values or having your Encode function return false.

Having your Encode function be able to return false is useful, particularly when you think about how you're going to signal recursion exceeded (which your current encode functions don't yet handle). It can also be used to signal "unrepresentable" values, such as dates outside of this range.

The fun part about strings and dates is that you can actually easily compare if a date is < "1950" or > "2049" and you'll get the right results - because the ASCII ordinals themselves are less than or greater (e.g. "1830" is < "1950" in both string and numeral form, just like "2051" is > "2049". So you could do something as

```
std::string year = iso_date.substr(0, 4);
if (year < "1950" or year > "2049")
  return;  // or return false, if/when you support error return
```